lib/mayor_game/city_calculator/city_helpers.ex

```diff
@@ -378,100 +379,42 @@ defmodule MayorGame.CityHelpers do
-
-          unhoused_citizens =
-            if acc.housing_left <= 0 && citizen.age < 10000 && !pollution_death,
-              do: [citizen | acc.unhoused_citizens],
-              else: acc.unhoused_citizens
-
-          polluted_citizens =
-            if pollution_death && citizen.age < 10000,
-              do: [citizen | acc.polluted_citizens],
-              else: acc.polluted_citizens
-
-          old_citizens =
-            if citizen.age > 10000,
-              do: [citizen | acc.old_citizens],
-              else: acc.old_citizens
-
-          # spawn new citizens if conditions are right; age, random, housing exists
-          reproducing_citizens =
-            if citizen.age > 500 and citizen.age < 2000 and
-                 :rand.uniform(length(city_baked_details.citizens) + 1) == 1,
-               do: [citizen | acc.reproducing_citizens],
-               else: acc.reproducing_citizens
-
-          will_citizen_learn =
-            rem(world.day, 365) == 0 && citizen.education < 5 &&
-              citizen.education < 5 &&
-              acc.education_left[citizen.education + 1] > 0 && citizen.age < 10000 &&
-              !pollution_death
-
-          education_left =
-            if will_citizen_learn do
-              Map.update!(acc.education_left, citizen.education + 1, &(&1 - 1))
-            else
-              acc.education_left
-            end
-
-          educated_citizens =
-            if will_citizen_learn do
-              Map.update!(acc.educated_citizens, citizen.education + 1, &[citizen | &1])
-            else
-              acc.educated_citizens
-            end
-
-          # return
-          %{
-            housing_left: housing_left,
-            education_left: education_left,
-            educated_citizens: educated_citizens,
-            housed_unemployed_citizens: housed_unemployed_citizens,
-            unhoused_citizens: unhoused_citizens,
-            housed_employed_staying_citizens: housed_employed_staying_citizens,
-            housed_employed_looking_citizens: housed_employed_looking_citizens,
-            polluted_citizens: polluted_citizens,
-            old_citizens: old_citizens,
-            reproducing_citizens: reproducing_citizens
-          }
+        if !pollution_reached do # so much we can skip if we don't do this check every time
+          fn citizen, acc ->
+            tax_too_high = :rand.uniform() < :math.pow(city.tax_rates[to_string(citizen.education)], 6 - citizen.education)
+            employable = acc.housing_left > 0 && citizen.has_job && citizen.age < 10000
+            will_citizen_learn = time_to_learn && citizen.education < 5 && acc.education_left[citizen.education + 1] > 0 && citizen.age < 10000
+            
+            acc |> Map.update!(housing_left:, if acc.housing_left > 0 do &(&1 - 1) else identity end)
+            |> Map.update!(education_left:, if will_citizen_learn do Map.update!(acc.education_left, citizen.education + 1,  &(&1 - 1)) else identity end)
```

Comment:
I don't think elixir has an identity function built in (or at least, I've never used it) — is this just meant to be the existing value for that key otherwise?

Reply:
I thiiiink the way to do this would be 
```ex 
 |> Map.update!(:education_left, if will_citizen_learn do Map.update!(acc.education_left, citizen.education + 1,  &(&1 - 1)) else &(&1) end)
```

Reply:
I was seeing this one https://hexdocs.pm/elixir/master/Function.html#identity/1 but, I think that would do it, too!

Reply:
oh wow! TIL, i think this is new-ish. I should try to update my elixir version